### PR TITLE
NIFI-2728 - Attempt to fix travis-ci build woes

### PR DESCRIPTION
**Pull request by trixpan**

### CODE REVIEW

**Comment by trixpan** [comment]
@mcgilman 

can you please review this and if in agreement merge it to master?

Current travis-ci settings seem to be unable to reliably build nifi. I fiddled with Xmx settings, enabled caching and added a license to the .travis.yml file and it seems to have alleviated the situation.

**Comment by mcgilman** [comment]
Just quickly glancing, I think that the contrib-check profile has issues with multi-threaded builds. We've seen issues with it in the past.

**Comment by trixpan** [comment]
@mcgillman, what do you think about leaving the - T2.0C setting to speed up build and adding an extra post-compile step to run - Pcontrib-check single threaded (this is similar to spark's approach)?

**Comment by trixpan** [comment]
Seems like moving `mvn -T2.0C clean install` to travis install phase and `mvn checkstyle:check` to the script phase did the trick.

around 25% reduction in build time and a far more reliable build process than previously, while preserving  `-PContrib-check` single-threaded execution.

**Comment by mcgilman** [comment]
@trixpan I'm not sure we want to move away from the contrib-check profile as this also ran through the rat plugin which verified files contained license headers, etc.

@apiri Would like to get your input on this too. Thanks.

**Comment by trixpan** [comment]
Matt, the profile is still being run but it run after compilation. 

First (install) travis performs a multi threaded mvn build. Immediately aftet this completes it runs a mvn checkstyle:check with a single thread. 

This allows the best of both worlds: speedy compilation with ongoing style and licensing enforcement

**Comment by mcgilman** [comment]
Weird. Maybe my browser had cached something. When I looked at this last night the contrib-check profile wasn't there. I see it now.

Anyways, this looks promising just want to make sure I'm not missing something. Would like @apiri to verify as well.

**Comment by apiri** [comment]
Seems fair, although I would almost prefer that the two items get flipped in terms of testing.  Always run contrib check (without tests) during the build and then run the full test suite.  In the linked build, it seems the tests failure from the environment seemingly prevented the contrib check from running, which, as a reviewer, is typically the most time consuming part of the review process.

Thoughts?

**Comment by trixpan** [comment]
@apiri makes complete sense. afaik this can be achieved using 

`mvn apache-rat:check and checkstyle:check `

so I added the above command to the before_install stage of the travis job. 

This should "fast fail" jobs failing to comply with development guidelines and by consequence, reduce a bit the load on travis-ci infrastructure. Very good call.

File: .travis.yml [hunk 1]
```diff
@@ -6,12 +21,22 @@ os:
...
+
 # before_install aids in a couple workarounds for issues within the Travis-CI environment
 #   1. Workaround for buffer overflow issues with OpenJDK versions of java as per https://github.com/travis-ci/travis-ci/issues/5227#issuecomment-165135711
 #   2. Workaround for non-existent Maven repository as per: https://github.com/travis-ci/travis-ci/issues/4629#issuecomment-131214465
 before_install:
   - sudo hostname "$(hostname | cut -c1-63)"
   - sed -e "s/^\\(127\\.0\\.0\\.1.*\\)/\\1 $(hostname | cut -c1-63)/" /etc/hosts | sudo tee /etc/hosts
   - sed -i.bak -e 's|https://nexus.codehaus.org/snapshots/|https://oss.sonatype.org/content/repositories/codehaus-snapshots/|g' ~/.m2/settings.xml
+  - echo "MAVEN_OPTS='-Xmx2g -XX:+UseG1GC'" > ~/.mavenrc
+  - mvn -Pcontrib-check apache-rat:check checkstyle:check
```
Comment (apiri) [inline .travis.yml]: Hey @trixpan,
I am a bit confused as it seems to be redundant.  contrib-check is a profile which runs RAT and checkstyle.  Am I overlooking something?

Comment (trixpan) [inline .travis.yml]: @apiri 

Yeah... that was a bit weird but it was an attempted workaround run checkstyle:check before compilation time. 

Sadly it didn't work, when we run checkstyle check, it triggers some style errors that have been disabled in our style rules:

`[WARNING] src/main/java/org/apache/nifi/reporting/ReportingTask.java[64:8] (javadoc) JavadocMethod: Unable to get class information for @throws tag 'org.apache.nifi.reporting.InitializationException'.`

**Comment by trixpan** [comment]
@apiri, @mcgilman 

I did a few new tweaks so that job sequence is:

1. rat check (before_install)
2. parallel compile with skipTests (install)
3. style check and test via mvn verify (script)

As mentioned above previously I could not find a way of running `checkstyle:check` before compilation, while respecting our checkstyle rules so to prevent errors such as:

`[WARNING] src/main/java/org/apache/nifi/reporting/ReportingTask.java[64:8] (javadoc) JavadocMethod: Unable to get class information for @throws tag 'org.apache.nifi.reporting.InitializationException'.`

**Comment by trixpan** [comment]
@ijokarumawak

**Comment by mcgilman** [comment]
Thanks @trixpan! This has been merged to master. Sorry it took a few days but wanted to make sure that others that were more familiar with our CI had a chance to review.

**Comment by trixpan** [comment]
@mcgilman would you mind pushing this commit to 0.x as well?

**Comment by mcgilman** [comment]
@trixpan The patch does not apply cleanly to 0.x. Would you mind posting another PR for that branch specifically.

**Comment by trixpan** [comment]
@mcgilman will do

**Pull request closed without merging**